registry.js

```diff
@@ -0,0 +1,134 @@
+class CustomAttributeRegistry {
+  constructor(ownerDocument){
+    if(!ownerDocument) {
+      throw new Error("Must be given a document");
+    }
+
+    this.ownerDocument = ownerDocument;
+    this._attrMap = new Map();
+    this._elementMap = new WeakMap();
+    this._observe();
+  }
+
+  define(attrName, Constructor) {
+    this._attrMap.set(attrName, Constructor);
+    this._upgradeAttr(attrName);
+  }
+
+  get(element, attrName) {
+    var map = this._elementMap.get(element);
+    if(!map) return;
+    return map.get(attrName);
+  }
+
+  _getConstructor(attrName){
+    return this._attrMap.get(attrName);
+  }
+
+  _observe(){
+    var customAttributes = this;
+    var root = this.ownerDocument;
+    var downgrade = this._downgrade.bind(this);
+    var upgrade = this._upgradeElement.bind(this);
+
+    this.observer = new MutationObserver(function(mutations){
+      forEach.call(mutations, function(m){
+        if(m.type === 'attributes') {
+          var attr = customAttributes._getConstructor(m.attributeName);
+          if(attr) {
+            customAttributes._found(m.attributeName, m.target, m.oldValue);
+          }
+        }
+        // chlidList
+        else {
+          forEach.call(m.removedNodes, downgrade);
+          forEach.call(m.addedNodes, upgrade);
```

Comment:
`forEach` is now undefined - think it was at the top of the file? Also I'm pretty sure `mutations` passed to the observer callback is already an Array so should only need the `forEach.call` on the above two lines.